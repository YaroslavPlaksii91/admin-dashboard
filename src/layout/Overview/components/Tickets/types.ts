export type TicketsItemType = {
  title: string;
  count: number;
};

export type TicketsProps = {
  data: TicketsItemType[];
};
