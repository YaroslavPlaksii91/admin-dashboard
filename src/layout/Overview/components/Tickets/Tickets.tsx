import { FC } from 'react';
import { List, ListItem, Typography, Box } from '@mui/material';

import { Link } from '@components/Link/Link';

import { TicketsProps } from './types';

export const Tickets: FC<TicketsProps> = ({ data }) => {
  return (
    <Box
      sx={{
        width: '50%',
        padding: '32px 0 8px',
        border: '1px solid',
        borderColor: 'borderColor',
        borderRadius: '8px',
        backgroundColor: 'mainWhiteColor',
      }}
    >
      <Box
        sx={{
          display: 'flex',
          justifyContent: 'space-between',
          mb: '8px',
          padding: '0 32px',
        }}
      >
        <Typography variant="h3">Unresolved tickets</Typography>
        <Link url="#">View details</Link>
      </Box>

      <Typography
        sx={{
          fontSize: '12px',
          lineHeight: 1.3,
          letterSpacing: 0.1,
          mb: 4,
          padding: '0 32px',
          color: 'secondaryTextColor',
        }}
      >
        Group:{' '}
        <Typography
          component="span"
          sx={{
            fontSize: '12px',
            fontWeight: 600,
            lineHeight: 1.3,
            letterSpacing: 0.1,
            color: 'grayDarkColor',
          }}
        >
          Support
        </Typography>
      </Typography>

      <List>
        {data.map(item => (
          <ListItem
            key={item.title}
            sx={{
              display: 'flex',
              justifyContent: 'space-between',
              padding: '20px 32px 18px',
              '&:not(:last-child)': {
                borderBottom: '1px solid',
                borderBottomColor: 'borderColor',
              },
            }}
          >
            <Typography
              sx={{
                fontSize: '14px',
                fontWeight: 600,
                lineHeight: 1.3,
                letterSpacing: 0.2,
              }}
            >
              {item.title}
            </Typography>
            <Typography
              component="span"
              sx={{
                fontSize: '14px',
                fontWeight: 600,
                lineHeight: 1.3,
                letterSpacing: 0.2,
                color: 'secondaryTextColor',
              }}
            >
              {item.count}
            </Typography>
          </ListItem>
        ))}
      </List>
    </Box>
  );
};
