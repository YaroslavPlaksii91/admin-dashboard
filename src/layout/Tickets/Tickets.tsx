import { FC, useState, useEffect } from 'react';
import { Box } from '@mui/material';

import { Pagination } from '@components/Pagination/Pagination';
import { ActionButtons } from '@components/ActionButtons/ActionButtons';
import { usePagination } from '@components/Pagination/usePagination';
import { getTickets } from '@services/getTickets';

import { TicketsItem } from './components/TicketsItem/TicketsItem';
import { TicketType } from './components/TicketsItem/types';
import { Heading } from './components/Heading/Heading';

export const Tickets: FC = () => {
  const [tickets, setTickets] = useState<TicketType[] | []>([]);

  const { getCurrentPageData, page, setPage, rowsPerPage, setRowsPerPage } =
    usePagination(tickets);

  useEffect(() => {
    const fetchData = async () => {
      const data = await getTickets();

      setTickets(data);
    };

    fetchData();
  }, []);

  const onSortClick = () => {
    console.log('Sort');
  };

  const onFilterClick = () => {
    console.log('Filter');
  };

  const onAddClick = () => {
    console.log('Add ticket');
  };

  return (
    <Box
      sx={{
        height: '100%',
        pt: '37px',
        border: '1px solid',
        borderColor: 'borderColor',
        borderRadius: '8px',
        backgroundColor: 'mainWhiteColor',
      }}
    >
      <ActionButtons
        addButtonName="Add ticket"
        onAddClick={onAddClick}
        onFilterClick={onFilterClick}
        onSortClick={onSortClick}
      />

      <Heading />

      {getCurrentPageData().map((ticket: any) => (
        <TicketsItem ticket={ticket} key={ticket.id} />
      ))}

      <Pagination
        count={tickets.length}
        page={page}
        setPage={setPage}
        rowsPerPage={rowsPerPage}
        setRowsPerPage={setRowsPerPage}
      />
    </Box>
  );
};
