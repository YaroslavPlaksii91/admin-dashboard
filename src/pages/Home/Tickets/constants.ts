export const TICKETS_COLUMNS = Object.freeze([
  {
    title: 'Ticket details',
    name: 'title',
    size: 5,
  },
  {
    title: 'Customer name',
    name: 'customerName',
    size: 3,
  },
  {
    title: 'Date',
    name: 'date',
    size: 2,
  },
  {
    title: 'Priority',
    name: 'priority',
    size: 2,
  },
]);

export const PRIORITY_OPTIONS = Object.freeze(['high', 'low', 'normal']);
