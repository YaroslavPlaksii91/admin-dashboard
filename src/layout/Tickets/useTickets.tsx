import { useState, useEffect } from 'react';

import { getTickets } from '@services/getTickets';
import { ROWS_PER_PAGE } from '@components/Pagination/constants';

import { TicketType } from './components/TicketsItem/types';

export const useTickets = () => {
  const [tickets, setTickets] = useState<TicketType[] | []>([]);
  const [page, setPage] = useState(0);

  useEffect(() => {
    const fetchData = async () => {
      const data = await getTickets();

      setTickets(data);
    };

    fetchData();
  }, []);

  const getCurrentPageTickets = () => {
    const startIndex = page * ROWS_PER_PAGE;
    const endIndex = startIndex + ROWS_PER_PAGE;

    return tickets.slice(startIndex, endIndex);
  };

  const ticketsCount = tickets.length;

  const onSortClick = () => {
    console.log('Sort');
  };

  const onFilterClick = () => {
    console.log('Filter');
  };

  const onAddClick = () => {
    console.log('Add ticket');
  };

  return {
    getCurrentPageTickets,
    page,
    setPage,
    onSortClick,
    onFilterClick,
    onAddClick,
    ticketsCount,
  };
};
