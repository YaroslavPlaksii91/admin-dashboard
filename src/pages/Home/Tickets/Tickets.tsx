import { FC, useState, useEffect } from 'react';
import { Box } from '@mui/material';

import { Pagination } from '@components/Pagination/Pagination';
import { usePagination } from '@components/Pagination/usePagination';
import { ModalComponent } from '@components/Modal/Modal';
import {
  getTickets,
  createTicket,
  deleteTicket,
  editTicket,
} from '@services/db/tickets';
import { useSort } from '@hooks/useSort';
import { useFilter } from '@hooks/useFilter';

import { TicketsItem } from './components/TicketsItem/TicketsItem';
import { TicketType } from './components/TicketsItem/types';
import { TicketsForm } from './components/TicketsForm/TicketsForm';
import { TicketsFormData } from './components/TicketsForm/types';
import { TICKETS_COLUMNS } from './constants';
import { Heading } from '../components/Heading/Heading';
import { ActionButtons } from '../components/ActionButtons/ActionButtons';

export const Tickets: FC = () => {
  const [tickets, setTickets] = useState<TicketType[]>([]);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [filterValue, setFilterValue] = useState('');
  const [activeTicketId, setActiveTicketId] = useState('');

  const filteredData = useFilter(tickets, filterValue, 'priority');

  const { sortedData, handleSort, sortKey, sortDirection } =
    useSort(filteredData);

  const { getCurrentPageData, page, setPage, rowsPerPage, setRowsPerPage } =
    usePagination(sortedData);

  const sortOptions = TICKETS_COLUMNS.map(col => col.name);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const data = await getTickets();
        data && setTickets(data);
      } catch (error) {
        console.error('Error while fetching data: ', error);
      }
    };

    fetchData();
  }, []);

  const onAddButtonClick = () => {
    setActiveTicketId('');
    setIsModalOpen(true);
  };

  const onEditButtonClick = (id: string) => {
    setIsModalOpen(true);
    setActiveTicketId(id);
  };

  const handleAddTicket = async ({
    title,
    date: { $d },
    name,
    priority,
  }: TicketsFormData) => {
    const newTicket = {
      id: String(tickets.length + 1),
      title,
      customerName: name,
      customerDate: new Date().toISOString(),
      date: new Date($d).toISOString(),
      priority,
      image:
        'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/574.jpg',
      updated: new Date().toISOString(),
    };

    const createdTicket = await createTicket(newTicket);

    setTickets([createdTicket, ...tickets]);

    setIsModalOpen(false);
  };

  const handleEditTicket = async ({
    title,
    date: { $d },
    name,
    priority,
  }: TicketsFormData) => {
    const body = {
      title,
      customerName: name,
      customerDate: new Date().toISOString(),
      date: new Date($d).toISOString(),
      priority,
      image:
        'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/574.jpg',
      updated: new Date().toISOString(),
    };

    const updatedTicket = await editTicket(activeTicketId, body);

    const newTickets = tickets.map(ticket =>
      ticket.id === activeTicketId ? updatedTicket : ticket,
    );

    setTickets([...newTickets]);

    setIsModalOpen(false);
  };

  const handleDeleteTicket = async (id: string) => {
    await deleteTicket(id);

    setTickets([...tickets.filter(ticket => ticket.id !== id)]);
  };

  return (
    <Box
      sx={{
        height: '100%',
        pt: '37px',
        border: '1px solid',
        borderColor: 'borderColor',
        borderRadius: '8px',
        backgroundColor: 'mainWhiteColor',
      }}
    >
      <ActionButtons
        addButtonName="Add ticket"
        onAddClick={onAddButtonClick}
        handleSort={handleSort}
        sortOptions={sortOptions}
        filterValue={filterValue}
        setFilterValue={setFilterValue}
        filterTitle="Priority"
        filterOptions={['high', 'normal', 'low']}
      />

      <Heading
        handleSort={handleSort}
        sortKey={sortKey}
        sortDirection={sortDirection}
        columns={TICKETS_COLUMNS}
      />

      {getCurrentPageData().map(ticket => (
        <TicketsItem
          ticket={ticket}
          key={ticket.id}
          handleDelete={() => handleDeleteTicket(ticket.id)}
          handleEdit={() => onEditButtonClick(ticket.id)}
        />
      ))}

      <Pagination
        count={filteredData.length}
        page={page}
        setPage={setPage}
        rowsPerPage={rowsPerPage}
        setRowsPerPage={setRowsPerPage}
      />

      <ModalComponent
        isOpen={isModalOpen}
        title="Add tickets"
        onClose={() => setIsModalOpen(false)}
      >
        <TicketsForm
          addTicket={handleAddTicket}
          editTicket={handleEditTicket}
          activeTicket={tickets.find(ticket => ticket.id === activeTicketId)}
        />
      </ModalComponent>
    </Box>
  );
};
