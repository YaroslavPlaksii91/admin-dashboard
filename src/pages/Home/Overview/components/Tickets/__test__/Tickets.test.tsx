import { cleanup, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Tickets } from '../Tickets';

describe('Tickets component', () => {
  afterEach(cleanup);

  const mockData = [
    { title: 'Ticket 1', count: 3 },
    { title: 'Ticket 2', count: 5 },
    { title: 'Ticket 3', count: 2 },
  ];

  it('should be rendered', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Tickets data={mockData} />
      </BrowserRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the list of tickets', () => {
    render(
      <BrowserRouter>
        <Tickets data={mockData} />
      </BrowserRouter>,
    );

    mockData.forEach(item => {
      expect(screen.getByText(item.title)).toBeInTheDocument();
      expect(screen.getByText(item.count.toString())).toBeInTheDocument();
    });
  });

  it('displays "Unresolved tickets" heading', () => {
    render(
      <BrowserRouter>
        <Tickets data={mockData} />
      </BrowserRouter>,
    );

    expect(screen.getByText('Unresolved tickets')).toBeInTheDocument();
  });

  it('displays "View details" link', () => {
    render(
      <BrowserRouter>
        <Tickets data={mockData} />
      </BrowserRouter>,
    );

    expect(screen.getByText('View details')).toBeInTheDocument();
  });

  it('displays the group name', () => {
    render(
      <BrowserRouter>
        <Tickets data={mockData} />
      </BrowserRouter>,
    );

    expect(screen.getByText('Group:')).toBeInTheDocument();
    expect(screen.getByText('Support')).toBeInTheDocument();
  });
});
