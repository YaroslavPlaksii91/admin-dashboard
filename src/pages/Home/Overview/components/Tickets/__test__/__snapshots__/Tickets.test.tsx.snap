// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tickets component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1wjthba"
  >
    <div
      class="MuiBox-root css-tgxzo7"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
      >
        Unresolved tickets
      </h3>
      <a
        href="/"
      >
        View details
      </a>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1u7yd7a-MuiTypography-root"
    >
      Group: 
      <span
        class="MuiTypography-root MuiTypography-body1 css-u5hitt-MuiTypography-root"
      >
        Support
      </span>
    </p>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-pcq8dq-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-b34im6-MuiTypography-root"
        >
          Ticket 1
        </p>
        <span
          class="MuiTypography-root MuiTypography-body1 css-1wgps9j-MuiTypography-root"
        >
          3
        </span>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-pcq8dq-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-b34im6-MuiTypography-root"
        >
          Ticket 2
        </p>
        <span
          class="MuiTypography-root MuiTypography-body1 css-1wgps9j-MuiTypography-root"
        >
          5
        </span>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-pcq8dq-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-b34im6-MuiTypography-root"
        >
          Ticket 3
        </p>
        <span
          class="MuiTypography-root MuiTypography-body1 css-1wgps9j-MuiTypography-root"
        >
          2
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
